Latest setuptools and flit now support PEP 639

not-my-profile opened the pull request:
This updates the now outdated notes added in
faf4c7d882ef2be310d5dc46d54ca666aa18821c.

Closes #1831.

<!-- readthedocs-preview python-packaging-user-guide start -->
----
📚 Documentation preview 📚:
* https://python-packaging-user-guide--1837.org.readthedocs.build/en/1837/guides/writing-pyproject-toml/
* https://python-packaging-user-guide--1837.org.readthedocs.build/en/1837/guides/licensing-examples-and-user-scenarios/
* https://python-packaging-user-guide--1837.org.readthedocs.build/en/1837/tutorials/packaging-projects/

<!-- readthedocs-preview python-packaging-user-guide end -->

not-my-profile commented on source/tutorials/packaging-projects.rst:
```diff
@@ -175,6 +175,9 @@ Frontends usually run builds in isolated environments, so omitting dependencies
 dependencies.
+The minimum version specified in the above code block is the one that introduced support
+for `license-files` being a list, if you need to use an older version you can adapt
+this requirement but will also need to specify the license files in a legacy format.
```
It would be neat to link the *Legacy license declaration* section from writing-pyproject-toml.rst here ... how do I do that in RST?

webknjaz replied:
If there's a label in that place, use the `:ref:` role. In general, you can see all linkable objects using intersphinx, for example.

webknjaz replied:
You'd have to add a label here https://github.com/pypa/packaging.python.org/blob/99469d7/source/guides/writing-pyproject-toml.rst?plain=1#L365-L367 to be able to reference it.

not-my-profile replied:
Thanks, I added a link to the relevant section of  writing-pyproject-toml.rst.

webknjaz commented:
Henry mentioned there's some limitations in Flit still. Perhaps, this should be documented.

cdce8p commented:
> Henry mentioned there's some limitations in Flit still.

Flitt `3.12` was released today with support for compound license expressions. Only thing missing now is `WITH` support. https://github.com/pypa/flit/issues/735 Though I suspect that won't be relevant to most people.

https://flit.pypa.io/en/stable/history.html#version-3-12

--
The PR mentions that "all" recommended backends now support PEP 639. I'm not sure about poetry-core. Yes, they accept the new syntax for `project.license` but I don't believe they have implemented any logic for it yet. AFAICT the metadata-version is also still kept at `2.3`.

https://github.com/python-poetry/poetry/issues/9670

--
As an aside, I did wonder if it would make sense to maintain a compatibility table for build-backends somewhere. I.e. a place that list you should use setuptools `>=77.0.3`, flit `>=3.12`, hatchling `>=1.27.0`, ... for support of PEP 639. There might be other major feature for which this might be relevant too. The alternative of course is to just always recommend the lastest release whenever possible.

not-my-profile commented:
> As an aside, I did wonder if it would make sense to maintain a compatibility table for build-backends somewhere.

I think that's a good idea. I now introduced a joined section [*license and license-files*](https://python-packaging-user-guide--1837.org.readthedocs.build/en/1837/guides/writing-pyproject-toml/#license-and-license-files) on the *Writing your pyproject.toml* page where I am now explaining the situation with the PEP and that there was a different format that's now deprecated. Then I have added a table listing build backend versions that introduced PEP 639 support. I think this should clear up user counfusion about why a build backend is complaining that `license` should be a dict.

I also went ahead and removed the *Legacy license declaration* section.

> The PR mentions that "all" recommended backends now support PEP 639. I'm not sure about poetry-core. Yes, they accept the new syntax for project.license but I don't believe they have implemented any logic for it yet. AFAICT the metadata-version is also still kept at 2.3.

Thanks for pointing this out! I tried the latest version of poetry and can confirm that `poetry build` still generates `Metadata-Version: 2.3`. I removed the "all" sentence and added poetry-core to the new version table with *not yet* linked to the relevant poetry issue.

>Flit-core 3.12 was released today with support for compound license expressions. Only thing missing now is WITH support. Though I suspect that won't be relevant to most people.

Great, I bumped the minimum version for flit-core to 3.12. Below the code block the aforementioned new *license and license-files* section is now linked for more context. I also added a footnote mentioning the lacking WITH support and linking the relevant flit issue.

cdce8p commented on source/guides/writing-pyproject-toml.rst:
```diff
@@ -322,22 +322,43 @@ You can also specify the format explicitly, like this:
+     - pdm-backend
+     - poetry-core
+   * - 1.27.0
+     - 77.0.1
```
```suggestion
     - 77.0.3
```
I'd suggest `77.0.3` here as it includes a few bug fixes for the initial implementation.

not-my-profile replied:
Changed :)

not-my-profile replied:
(I also added a respective footnote to the version table, linking the setuptools changelog.)

cdce8p commented on source/tutorials/packaging-projects.rst:
```diff
@@ -141,31 +141,31 @@ for more details.
 
         [build-system]
-        requires = ["setuptools>=61.0"]
+        requires = ["setuptools>=77.0.1"]
```
```suggestion
        requires = ["setuptools>=77.0.3"]
```

not-my-profile replied:
Changed :)

cdce8p commented on source/guides/writing-pyproject-toml.rst:
```diff
@@ -363,41 +384,11 @@ The custom identifiers must follow the SPDX specification,
     [project]
     license = "LicenseRef-My-Custom-License"
 
-Legacy license declaration
-''''''''''''''''''''''''''
-
-This can take two forms. You can put your license in a file, typically
-:file:`LICENSE` or :file:`LICENSE.txt`, and link that file here:
-
-.. code-block:: toml
-
-    [project]
-    license = {file = "LICENSE"}
-
-or you can write the name of the license:
-
-.. code-block:: toml
-
-    [project]
-    license = {text = "MIT License"}
-
-If you are using a standard, well-known license, it is not necessary to use this
-field. Instead, you should use one of the :ref:`classifiers` starting with ``License
-::``. (As a general rule, it is a good idea to use a standard, well-known
-license, both to avoid confusion and because some organizations avoid software
-whose license is unapproved.)
-
```
Even with support for PEP 639 in almost all build backends, I'm unsure if that should really be removed. The packaging ecosystems moves quite slow. It might be helpful to understand old configs.

Adding a warning here (if possible) might be the better choice.

not-my-profile replied:
You're right that we should still help users learn about the legacy format but I think referencing the respective section of PEP 621 for that is enough. I have added [a Note in the license section](https://python-packaging-user-guide--1837.org.readthedocs.build/en/1837/guides/writing-pyproject-toml/#license) saying:

> If you get a build error that license should be a dict/table, your build backend doesn’t yet support the new format. See the [above section](https://python-packaging-user-guide--1837.org.readthedocs.build/en/1837/guides/writing-pyproject-toml/#license-and-license-files) for more context. The now deprecated format is [described in PEP 621](https://peps.python.org/pep-0621/#license).

not-my-profile commented:
(Since I found out that the build backend tabs are present on two pages I have move them to a separate file and included them in both pages so that the minimum versions we specify there don't drift in the future.)

cdce8p commented on source/guides/writing-pyproject-toml.rst:
```diff
@@ -610,15 +575,23 @@ A full example
 
+.. [#flit-core-pep639] flit-core `does not yet <flit-issue-735_>`_ support WITH in SPDX license expressions.
+
+.. [#setuptools-pep639] setuptools 77.0.1 introduced support for :pep:`639`
+   but `77.0.3 fixed some related bugs <setuptools-changelog-77.0.3_>`_.
```
I'd just remove the footnote. In a few months it will be irrelevant that support was technically added with `77.0.1`.

not-my-profile replied:
Fair enough.

cdce8p reviewed (approved):
LGTM

AA-Turner commented on source/shared/build-backend-tabs.rst:
```diff
@@ -0,0 +1,34 @@
+    .. code-block:: toml
+
+        [build-system]
+        requires = ["flit_core >= 3.12.0"]
```
Flit has always included a `<4` in its own documentation:
```suggestion
        requires = ["flit_core>=3.12,<4"]
```

not-my-profile replied:
Added.

webknjaz reviewed (approved):
Thank you!
